### ISSUE TITLE
Test implementation of nats-server throws slow consumer with two connections

### ISSUE THREAD

**Issue by Kaspermn33** [nats-io/nats-server#3545]
## Defect
Hi, I've been creating a test, for a wrapper over Ubers zap library, that publishes onto NATS. The code shown below, is a minimal example, mocking the wrapper, but still causing the test to fail.

The test originally failed because the `select` timed out before the channel received any data, increasing the timeout does nothing, other than increasing the time before it fails, it will still fail at 1 minute. 

After debugging, I found that the server threw a `slow consumer` error, which you can also see in the actual output, I don't quite understand why it would do that, since only 1 message is published.

Whether the test passes or fails is very inconsistent. For me it failed on the fifth attempt, but it's much more consistent on GitHub Workflows.

I'm not sure whether I'm doing something wrong, or if there's a bug inside of the test implementation of the server?

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [ ] Included `nats-server -DV` output
N/A
It's the test implementation of the server.
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats-server` and affected client libraries used:

N/A 

#### OS/Container environment:
macOS Monterey version 12.6
&
GitHub Workflows `runs-on: ubuntu-latest`

#### Steps or code to reproduce the issue:

Code snippet of entire go file.

Run of go test `go test ./... -v --race`

```golang
package minimal

import (
	"fmt"
	"testing"
	"time"

	"github.com/nats-io/nats-server/v2/server"
	natsserver "github.com/nats-io/nats-server/v2/test"
	"github.com/nats-io/nats.go"
)

func runServerOnPort(port int) *server.Server {
	opts := natsserver.DefaultTestOptions
	opts.Port = port
	return runServerWithOptions(&opts)
}

func runServerWithOptions(opts *server.Options) *server.Server {
	return natsserver.RunServer(opts)
}

func mockPublishMessage(subject, url string, data []byte) {
	var conn *nats.Conn
	var err error

	conn, err = nats.Connect(url)

	if err != nil {
		fmt.Println(err)
	}

	err = conn.Publish(subject, data)
	if err != nil {
		fmt.Println(err)
	}
	conn.Close()
}

func TestMinimalTest(t *testing.T) {
	tests := []struct {
		name    string
		subject string
		message string
	}{
		{name: "case1", subject: "some.subject", message: "some.message"},
		{name: "case2", subject: "some.subject", message: "some.message"},
		{name: "case3", subject: "some.subject", message: "some.message"},
		{name: "case4", subject: "some.subject", message: "some.message"},
		{name: "case5", subject: "some.subject", message: "some.message"},
		{name: "case6", subject: "some.subject", message: "some.message"},
		{name: "case7", subject: "some.subject", message: "some.message"},
		{name: "case8", subject: "some.subject", message: "some.message"},
		{name: "case9", subject: "some.subject", message: "some.message"},
		{name: "case10", subject: "some.subject", message: "some.message"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			natsServ := runServerOnPort(8484)
			defer natsServ.Shutdown()

			sURL := fmt.Sprintf("nats://127.0.0.1:%d", 8484)
			nc, err := nats.Connect(sURL, nats.ErrorHandler(func(nc *nats.Conn, sub *nats.Subscription, natsErr error) {
				fmt.Printf("error: %v\n", natsErr)
				if natsErr == nats.ErrSlowConsumer {
					t.Error("subscriber was marked as a slow consumer")
				}
			}))
			if err != nil {
				t.Error(err)
			}

			natsChan := make(chan *nats.Msg)
			defer close(natsChan)
			subscription, err := nc.ChanSubscribe(">", natsChan)

			defer func() {
				err = subscription.Unsubscribe()
				if err != nil {
					t.Error(err)
				}
			}()

			mockPublishMessage(tt.subject, sURL, []byte(tt.message))

			select {
			case msg := <-natsChan:
				msgString := string(msg.Data)
				if msgString != tt.message {
					t.Errorf("received NATS message not as expected, got %v, expected %v", msgString, tt.message)
				}
			case <-time.After(1 * time.Second):
				t.Error("no message received before timeout")
			}
		})
	}
}

```

#### Expected result:
=== RUN   TestMinimalTest
=== RUN   TestMinimalTest/case1
=== RUN   TestMinimalTest/case2
=== RUN   TestMinimalTest/case3
=== RUN   TestMinimalTest/case4
=== RUN   TestMinimalTest/case5
=== RUN   TestMinimalTest/case6
=== RUN   TestMinimalTest/case7
=== RUN   TestMinimalTest/case8
=== RUN   TestMinimalTest/case9
=== RUN   TestMinimalTest/case10
--- PASS: TestMinimalTest (0.31s)
    --- PASS: TestMinimalTest/case1 (0.03s)
    --- PASS: TestMinimalTest/case2 (0.03s)
    --- PASS: TestMinimalTest/case3 (0.03s)
    --- PASS: TestMinimalTest/case4 (0.03s)
    --- PASS: TestMinimalTest/case5 (0.03s)
    --- PASS: TestMinimalTest/case6 (0.03s)
    --- PASS: TestMinimalTest/case7 (0.04s)
    --- PASS: TestMinimalTest/case8 (0.03s)
    --- PASS: TestMinimalTest/case9 (0.03s)
    --- PASS: TestMinimalTest/case10 (0.03s)
PASS
ok  	github.com/minimal	0.515s

#### Actual result:
=== RUN   TestMinimalTest
=== RUN   TestMinimalTest/case1
=== RUN   TestMinimalTest/case2
=== RUN   TestMinimalTest/case3
=== RUN   TestMinimalTest/case4
=== RUN   TestMinimalTest/case5
=== RUN   TestMinimalTest/case6
=== RUN   TestMinimalTest/case7
error: nats: slow consumer, messages dropped
    minimal_test.go:67: subscriber was marked as a slow consumer
    minimal_test.go:94: no message received before timeout
=== RUN   TestMinimalTest/case8
=== RUN   TestMinimalTest/case9
=== RUN   TestMinimalTest/case10
--- FAIL: TestMinimalTest (1.31s)
    --- PASS: TestMinimalTest/case1 (0.04s)
    --- PASS: TestMinimalTest/case2 (0.03s)
    --- PASS: TestMinimalTest/case3 (0.03s)
    --- PASS: TestMinimalTest/case4 (0.03s)
    --- PASS: TestMinimalTest/case5 (0.03s)
    --- PASS: TestMinimalTest/case6 (0.03s)
    --- FAIL: TestMinimalTest/case7 (1.03s)
    --- PASS: TestMinimalTest/case8 (0.03s)
    --- PASS: TestMinimalTest/case9 (0.03s)
    --- PASS: TestMinimalTest/case10 (0.03s)
FAIL
FAIL	github.com/minimal	1.496s

**Comment by derekcollison** [comment]
The client will expect the subscriber to be able to handle messages delivered by the server. In the callback scenario that is set to an internal capacity IIRC of 65k message pending internally to the client.

When you use a channel and pass that to the subscriber that becomes the bounding limit, and if the clients internal processing of messages from the server can not place a message on the channel it will cause an error.

In your case you do a blocking channel, meaning unless someone is actively waiting on that channel to read from it the push (ch <- msg) call will fail and cause the error.

Try this..

```go
natsChan := make(chan *nats.Msg, 1024)
```

**Comment by Kaspermn33** [comment]
Thanks, that seems to have fixed the issue with the slow consumer.

But another issue has come up, where it just times out, without logging or printing anything else. 

I haven't been able to replicate this on my mac, but it happens on GitHub Workflows.

=== RUN   TestMinimalTest
=== RUN   TestMinimalTest/case1
    minimal_test.go:94: no message received before timeout
=== RUN   TestMinimalTest/case2
=== RUN   TestMinimalTest/case3
=== RUN   TestMinimalTest/case4
=== RUN   TestMinimalTest/case5
=== RUN   TestMinimalTest/case6
=== RUN   TestMinimalTest/case7
=== RUN   TestMinimalTest/case8
=== RUN   TestMinimalTest/case9
=== RUN   TestMinimalTest/case10
--- FAIL: TestMinimalTest (1.30s)
    --- FAIL: TestMinimalTest/case1 (1.04s)
    --- PASS: TestMinimalTest/case2 (0.01s)
    --- PASS: TestMinimalTest/case3 (0.03s)
    --- PASS: TestMinimalTest/case4 (0.03s)
    --- PASS: TestMinimalTest/case5 (0.03s)
    --- PASS: TestMinimalTest/case6 (0.04s)
    --- PASS: TestMinimalTest/case7 (0.03s)
    --- PASS: TestMinimalTest/case8 (0.03s)
    --- PASS: TestMinimalTest/case9 (0.03s)
    --- PASS: TestMinimalTest/case10 (0.03s)
FAIL
FAIL	github.com/DanaDynamics/backend/pkg/v0/minimal	1.320s
FAIL

**Comment by derekcollison** [comment]
I would not expect it to take long at all, but increase timeout to 2 or 3s and see if that resolves.

**Comment by Kaspermn33** [comment]
I increased the timeout to 3 seconds, but it still timed out. But it's incredibly inconsistent, and I needed to rerun the test 4 times before it decided to fail.

=== RUN   TestMinimalTest
=== RUN   TestMinimalTest/case1
=== RUN   TestMinimalTest/case2
=== RUN   TestMinimalTest/case3
=== RUN   TestMinimalTest/case4
    minimal_test.go:94: no message received before timeout
=== RUN   TestMinimalTest/case5
=== RUN   TestMinimalTest/case6
=== RUN   TestMinimalTest/case7
=== RUN   TestMinimalTest/case8
=== RUN   TestMinimalTest/case9
=== RUN   TestMinimalTest/case10
--- FAIL: TestMinimalTest (3.25s)
    --- PASS: TestMinimalTest/case1 (0.03s)
    --- PASS: TestMinimalTest/case2 (0.03s)
    --- PASS: TestMinimalTest/case3 (0.00s)
    --- FAIL: TestMinimalTest/case4 (3.01s)
    --- PASS: TestMinimalTest/case5 (0.03s)
    --- PASS: TestMinimalTest/case6 (0.03s)
    --- PASS: TestMinimalTest/case7 (0.03s)
    --- PASS: TestMinimalTest/case8 (0.03s)
    --- PASS: TestMinimalTest/case9 (0.03s)
    --- PASS: TestMinimalTest/case10 (0.03s)
FAIL

**Comment by kozlovic** [comment]
That is because your mockPublishMessage() function creates its own connection to publish the message. So there is no guarantee that the subscription has been processed by the server before the message is sent. Adding nc.Flush() after creating the subscription will solve that problem (PS: you get the error on nc.ChanSubscribe() but don't check it)

**Comment by Kaspermn33** [comment]
Thank you very much.
This solved my problem.

I must've missed the check, when I created the minimal example :grin:
